parsl/tests/test_staging/test_elaborate_noop_file.py

```diff
@@ -0,0 +1,82 @@
+# this replaces the no-op file staging provider with a more
+# complicated no-op provider that launches (empty) tasks
+# for staging
+
+# the elaborate no-op provider is based around the structure
+# of the globus staging provider
+
+import logging
+import pytest
+
+import parsl
+
+from parsl import bash_app, python_app
+from parsl.config import Config
+from parsl.data_provider.files import File
+from parsl.executors.threads import ThreadPoolExecutor
+from parsl.tests.test_staging.staging_provider import NoOpTestingFileStaging
+from parsl.tests.utils import get_rundir
+
+logger = logging.getLogger(__name__)
+
+
+@bash_app
+def touch(filename, inputs=[], outputs=[]):
+    return "touch {}".format(filename)
+
+
+@python_app
+def app_test_in(file):
+    # does not need to do anything as this
+    # is just for exercising staging
+    pass
+
+
+@pytest.mark.local
+def test_regression_stage_out_does_not_stage_in():
+    no_stageout_config = Config(
+        executors=[
+            ThreadPoolExecutor(
+                label='local_threads',
+                storage_access=[NoOpTestingFileStaging(allow_stage_in=False)]
+            )
+        ],
+        run_dir=get_rundir()
+    )
+
+    parsl.load(no_stageout_config)
+
+    # test with no staging
+    touch("test.1", outputs=[File("test.1")]).result()
+
+    # test with stage-out
+    touch("test.2", outputs=[File("test.2")]).result()
+
+    parsl.dfk().cleanup()
+    parsl.clear()
+
+
+@pytest.mark.local
+def test_regression_stage_in_does_not_stage_out():
+    no_stageout_config = Config(
+        executors=[
+            ThreadPoolExecutor(
+                label='local_threads',
+                storage_access=[NoOpTestingFileStaging(allow_stage_out=False)]
+            )
+        ],
+        run_dir=get_rundir()
+    )
+
+    parsl.load(no_stageout_config)
+
+    # TODO create a file not using a task
+
+    f = open("test.3", "a")
+    f.write("test")
+    f.close()
+
+    app_test_in(File("test.3")).result()
```

Comment:
Do we need an invocation that needs stage_out to raise an exception to test the alternate path?

Reply:
These test cases were intended to test that exceptions *don't* happen but I think it would make sense to test that exceptions from providers *do* propagate correctly too.

Reply:
I've added two test cases for that in 2ef6ca05d